pkg/addonmanager/controllers/agentdeploy/controller.go

```diff
@@ -237,44 +271,202 @@ func (c *addonDeployController) applyWork(ctx context.Context, required *workapi
+		} else {
+			deployManifests = append(deployManifests, workapiv1.Manifest{
+				RawExtension: runtime.RawExtension{Raw: rawObject},
+			})
+		}
+	}
+	deployWorkName := fmt.Sprintf("addon-%s-deploy", addonName)
+	deployManifestWork = newDeployManifestWork(deployWorkName, addonName, cluster, deployManifests)
+	if len(hookManifests) != 0 {
+		hookWorkName := fmt.Sprintf("addon-%s-pre-delete", addonName)
+		hookManifestWork = newHookManifestWork(hookWorkName, addonName, cluster, hookManifests, manifestConfigs)
+		return deployManifestWork, hookManifestWork, nil
+	}
+
+	return deployManifestWork, nil, nil
+}
+
+func ManifestsEqual(new, old []workapiv1.Manifest) bool {
+	if len(new) != len(old) {
+		return false
+	}
+
+	for i := range new {
+		if !equality.Semantic.DeepEqual(new[i].Raw, old[i].Raw) {
+			return false
+		}
+	}
+	return true
+}
+
+func newDeployManifestWork(workName, addonName, clusterName string, manifests []workapiv1.Manifest) *workapiv1.ManifestWork {
+	return &workapiv1.ManifestWork{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      fmt.Sprintf("addon-%s-deploy", addonName),
-			Namespace: cluster,
+			Name:      workName,
+			Namespace: clusterName,
 			Labels: map[string]string{
 				constants.AddonLabel: addonName,
 			},
 		},
 		Spec: workapiv1.ManifestWorkSpec{
 			Workload: workapiv1.ManifestsTemplate{
-				Manifests: []workapiv1.Manifest{},
+				Manifests: manifests,
 			},
 		},
 	}
+}
 
-	for _, object := range objects {
-		rawObject, err := runtime.Encode(unstructured.UnstructuredJSONScheme, object)
-		if err != nil {
-			return nil, err
+func newManifestConfigOption(obj *unstructured.Unstructured) workapiv1.ManifestConfigOption {
+	var resource string
+	var jsonPath workapiv1.JsonPath
+	gvk := obj.GroupVersionKind()
+	switch gvk.Kind {
+	case "Job":
+		resource = "jobs"
+		jsonPath = workapiv1.JsonPath{
+			Name: "isCompleted",
+			Path: `.status.conditions[?(@.type=="Complete")].status`,
 		}
-		work.Spec.Workload.Manifests = append(work.Spec.Workload.Manifests, workapiv1.Manifest{
-			RawExtension: runtime.RawExtension{Raw: rawObject},
-		})
+	case "Pod":
+		resource = "pods"
+		jsonPath = workapiv1.JsonPath{
+			Name: "isReady",
+			Path: `.status.conditions[?(@.type=="Ready")].status`,
+		}
+	default:
+		return workapiv1.ManifestConfigOption{}
+	}
+	return workapiv1.ManifestConfigOption{
+		ResourceIdentifier: workapiv1.ResourceIdentifier{
+			Group:     gvk.Group,
+			Resource:  resource,
+			Name:      obj.GetName(),
+			Namespace: obj.GetNamespace(),
+		},
+		FeedbackRules: []workapiv1.FeedbackRule{
+			{
+				Type:      workapiv1.JSONPathsType,
+				JsonPaths: []workapiv1.JsonPath{jsonPath},
+			},
+		},
+	}
+}
+func newHookManifestWork(
+	workName, addonName, clusterName string,
+	manifests []workapiv1.Manifest,
+	manifestConfigs []workapiv1.ManifestConfigOption) *workapiv1.ManifestWork {
+	hookWork := &workapiv1.ManifestWork{
```

Comment:
I think you can call newDeployManifestWork here.

Reply:
done